### DISCUSSION TITLE
Car charging

### DISCUSSION THREAD

**Discussion by pacemaker82** [springfall2008/batpred#750] [General]
As I understand it I can set up PredBat to use Agile and connect to my car charging, and have predbat manage the car charging?

I set the car battery size, % target and current % in. apps.yaml, and set number of cars to 1.

I've set up all the sensors (I think) but the one that actually triggers a charge doesnt seem to be set. For my Tesla I have a sensor : "switch.charger" which gets turned on to start charging. But I have no idea how to set that up in apps.yaml. 

Is that what needs to be set for "car_charging_planned" ? Or am I just misunderstanding what predbat can do here...

**Comment by springfall2008** [comment]
You need to create a HA automation that turns your car charger on and off when the Predbat slot toggles. I'd recommend that it should run every 5 minutes just in case a toggle is missed

**Reply by springfall2008** [reply]
From the documentation, perhaps should be clearer

Use an automation based on binary_sensor.predbat_car_charging_slot to control when your car charges

**Reply by pacemaker82** [reply]
thanks trefor - I ended up going IOG anyway as it seems a better deal.

Q on that - I have this in my plan:
<img width="476" alt="Screenshot 2024-02-12 at 12 46 47" src="https://github.com/springfall2008/batpred/assets/157808562/7708902d-5b0f-467c-9c6a-abf7623d8d20">

So I am discharging assuming I will charge at 14.30. I know the car will be gone by then so I assume the slots are no longer cheap. I know Predbat cant predict this, but wondering what the best approach is - dont plug for charge until I know I am home for good for the day, so as to avoid this kind of situation where I am discharging but cant recharge at cheap?

**Reply by springfall2008** [reply]
When you unplug the car Intelligent will take the slots away again (it takes say 10-15 minutes to notice).

You can also change the setting in Predbat to ignore IO slots when the car is unplugged provided you have the car plugged in sensor configured correctly.

**Reply by pacemaker82** [reply]
Thanks - Had an interesting experience this morning, and wondering if there is a bug or enhancement here.

I plugged my car in last night and was given a massive charge session (7.30pm thru to 8am) (it's new and was delivered with no charge). I had set it to 90% target in Octopus and I have a granny charger.. so worked out almost right...

By 06.30am this morning the car hit 90% and stopped charging, so I still had 3 half hour slots at the 7.5p rate, according to the PredBat plan, which I assume the octopus integration is telling it too. **The predbat plan continued to run from grid until the planned 8am end.**

However looking now at the octopus integration "completed dispatches" attribute in the intelligent dispatching sensor, I can see that **the cheap half hour windows stopped at 6.30am**, there are none recorded after that. So a few questions:

1) Seems ok behaviour for Octopus, they said they would give me till 8am, charged it to target early, so the extra 3 half hour slots were not charging, so they are not "completed dispatches" anymore, they were removed. Right?
2) Seems these dispatches that aren't "used" or no longer valid, won't disappear from their API until after the fact, as predbat plan still had them as active up until 8am.
3) If so, should predbat consider the "is charging" sensor to ensure the plan is accurate and against the right cheap or expensive rate? `If planned dispatch but not charging, then consider it an expensive rate`?

If this is already a supported case, should I be setting these in the apps.yaml:

  car_charging_now:
    - 're:switch.charger'

  car_charging_now_response:
    - 'on'

**Reply by pacemaker82** [reply]
@springfall2008 I guess the challenge with adding the car_charging_now in apps.yaml is that it also now thinks there is a cheap window if charging starts on my car (like when I plug it in). The plan fixes itself after I turn it off though.

**Reply by matknowles** [reply]
This is exactly my situation as well, except using a Zappi charger. Have you found a solution yet? I've tried messing around with the new watchlist feature as well trying to figure out an automation to get predbat to change the plan to remove additional charging despite the car being full. On my Zappi theres a sensor for plug status that shows completed when the car's done, but I'm so new to home assistant in general and am not yet able to get it working satisfactorily. I did note though that despite the car being charged, the octopus app still showed an additional cheap slot that didn't go away, as the car was still plugged in.

I thought of making an automation as suggested that does something when the zappi status is completed, but I'm not sure what would work well. Considered switching predbat to monitor mode to stop the plan, or changing octopus_intelligent_charging in the hopes it'd change the plan, but it doesn't seem to.

Also noted on another discussion 463 "How can predbat better understand when the car is charging" about using the cars SoC sensor to do this, but for my case (a Hyundai Kona), the sensor only updates very infrequently and so isn't usable for quickly changing the plan/ removing the extra slots.

**Reply by RobinCu** [reply]
The TL;DR is don't worry about the extra charge slots staying in Predbat if using the Zappi charger integration.

The exact way Octopus handles unused charging smart slots on charger integration seems to vary.  For the Zappi the completed dispatches are only recorded via the public API (visible in HA) if a charge is delivered to the car during that 30-minute period.

With the Zappi (charger integration rather than car integration), the extra off-peak rate applies to the whole house even if the car is completed and remains plugged in.  So even though they don't show in HA, Octopus has recorded it.  Several people have tested this theory, and I can confirm it is always the case.  My average p/kWh since November is around 7.6p/kWh, as Predbat uses all the slots, and the only time I draw from the grid at 30.6p/kWh is when the house load exceeds the inverter or a few moments it takes for BottleCapDave's integration to notice a change.  

Although the Kona's SOC is infrequently updated, my answer in #463 is worth trying, as it slightly improves the plan because it's not using the full Octopus "charge to add" percentage.  It doesn't impact the stop/start of the charging, as Octopus are controlling that via the Zappi.  If you still want to go down the automation route based on Zappi Completed status, I can give you some pointers, but I'm not sure it will help.

I hope that makes some sense, but feel free to let me know if it's still not working for you, as it's been working well for me.

Side note: Octopus are working on making all historical smart slots available via their public API.

**Reply by matknowles** [reply]
Oh that's great news, thanks for confirming. We only had our solar/batteries installed a few weeks ago and moved to the intelligent octopus tariff even sooner than that, so we haven't received a bill yet, hence me being especially unsure.

I had already done as you suggested in 463, but I wasn't sure if it was helpful owing to the slow update interval. I did look at reducing this interval, but apparently it can take its toll on the 12v battery, so haven't messed with it.

It all seems to work excellently otherwise; I was just concerned about that edge case where the car is full but extra slots are still allocated. Thanks for putting my mind at rest.

**Reply by pacemaker82** [reply]
I have to say I had the opposite experience though. 

Octopus had stated planned dispatches up to 0800 hrs. Car finished at 0630, the planned dispatches didnt update so predbat continued to run from grid (assuming it was cheap). Next day the completed dispatches were up to 0630, there was no evidence from octopus that the cheap rate continued until 0800 from their API.

Unless you're saying you've proved it some other way...?

**Reply by matknowles** [reply]
Oh OK, that's interesting as that's pretty much my exact experience. Plugged in in the afternoon, was given 2130-2200 then 2230 to 0800. Checked my phone at about 0530 and the zappi had finished about 0200, but the octopus app still showed a dispatch until 0800 (and predbat quite rightly kept the extra charge/discharge plan in place, as this was what octopus and hence the integration was reporting)
Unfortunately as I mentioned I still haven't had a bill, so can't say for sure whether I'll be billed at cheap rate for that time or not.

**Reply by pacemaker82** [reply]
yeah the slot plan often doesn't update from octopus, unless you physically disconnect the car. 

Ideally predbat should have logic

- IF octopus charge slot AND is extra slot (meaning not the usual 2330-0530 window)
     - IF car is charging OR charged within 30 min slot
         - then we are good and do normal predbat things
     - ELSE
         - predbat should assume its not an official cheap slot and act appropriately

**Reply by RobinCu** [reply]
As I say, you will find the extra slots are billed 7.5p/kWh.  Unless you have an Ohme charger in which case there is a separate HA integration that manages the interaction with Predbat.

if you really want to cancel the home battery charging behaviour (if the car charge has ended), you could set an automation to set Predbat to read-only if the charge completes but it really would mean missing out on the cheap slots.

**Reply by pacemaker82** [reply]
so does that mean the "completed_dispatches" from the Octopus API cannot be trusted as the actual billed 7.5p rate windows?

**Reply by RobinCu** [reply]
Correct, the public API only records the charge slots where the car has received charge.  Octopus’ own system records all.  It’s a similar reason why Octopus Compare apps etc. don’t work with IOC.

**Reply by matknowles** [reply]
So to clarify then, as long as the car accepts any charge at all, all dispatches will be billed at cheap rate? Additionally, if I were to plug in but the car is already fully charged, would I get any additional cheap rates at all?

Apologies for my confusion.

**Reply by RobinCu** [reply]
@matknowles yes, spot on.  If the car charges it’s 7.5p and if the charge completes, any remaining smart slots are billed at 7.5p.  Charge complete doesn’t need to be 100% SOC. I keep my in-car limit at 70% unless planning a long trip.

@pacemaker82 I might have added to the confusion by saying “correct” to your last reply (apologies I was dashing out the door and didn’t re-read before hitting reply). I meant that the Octopus API completed dispatches can’t be trusted as the complete picture but if you see a slot there it will definitely be 7.5p.  They only stay on the API for about 24 hours though, so others have discussed a way to write them to a files. The point is that the extra slots where the car didn’t charge are still 7.5p.  

I experienced all this confusion and unknowns. Screenshotting the Octopus app, etc for a few days. but after a couple of months of bills where 98-99% of usage is at 7.5p/kWh, you soon put your trust and just enjoy it.  You can also request an interim bill from Octopus to double check.  The best way is via Twitter/X.

To be clear though for anyone referring to this in the future, the above only applies to charger-based IOG interventions like the Zappi.  Ohme has their own mechanism which seems to be different. It does make it overly complicated and results in a lot of muddled information online.

**Reply by pacemaker82** [reply]
Thanks @RobinCu - yeah exactly - its the fear - I had the same issue today, the planned dispatches were saying charge till 7.30am, and it finished at 6.45, so predbat happily carried on up to 7.30 assuming it was cheap (sounds like assuming correctly though). That 7-7.30 slot never appeared as a completed dispatch, which also adds to the fear, because the octopus integration then treats it as an expensive slot and adds up its cost appropriately. Hopefully Octopus will sort this out!

**Comment by pacemaker82** [comment]
@RobinCu wanted to confirm, got my bill today and indeed you are right, if Octopus said they were giving you a slot and they didnt use it, they still charge you at the cheap rate. So the completed dispatches are only showing when the car _actually charged in a cheap slot_, not the cheap slots themselves historically. This also means that the accumulation cost (and previous day cost) from the Octopus Integration from BottlecapDave is only semi-accurate as he bases the cheap slots off completed_dispatches only. I wonder how predbat is affected by that inaccuracy @springfall2008 in terms of the costs that it quotes?
